File: 04.Simulation模拟/leetcode_2078_maxDistance.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int maxDistance(vector<int>& colors) {
        int n = colors.size();
        if (colors[0] != colors[n-1]){
            return n-1;
        }
        int res = 0;
        for (int i=0; i<n; i++){
            if (colors[i] != colors[n-1]){
                res = n-1-i;
                break;
            }
        }
        for (int i=n-1; i>=0; i--){
            if (colors[i] != colors[0]){
                res = max(res, i);
                break;
            }
        }
        return res;
    }
};

int main(){
    Solution s;
    vector<int> colors = {1,1,1,6,1,1,1};
    cout << s.maxDistance(colors) << endl;
}

File: 08.BrainTwist脑筋急转弯/leetcode_2078_maxDistance.py
from typing import List

class Solution:
    def maxDistance(self, colors: List[int]) -> int:
        n = len(colors)
        if colors[0] != colors[-1]:
            return n-1
        res = 0
        def calc(start, end, space):
            for i in range(start, end, space):
                if colors[i] != colors[0]:
                    return i
        return max(n-1-calc(1, n, 1), calc(n-2, -1, -1))

if __name__ == "__main__":
    s = Solution()
    colors = [1,1,1,6,1,1,1]
    print(s.maxDistance(colors))
